data/src/main/java/com/microsoft/azure/kusto/data/auth/TokenProviderBase.java

```diff
@@ -0,0 +1,104 @@
+    protected static final String ERROR_ACQUIRING_APPLICATION_ACCESS_TOKEN = "Error acquiring ApplicationAccessToken";
+    protected final Logger logger = LoggerFactory.getLogger(getClass());
+    protected static final int TIMEOUT_MS = 20 * 1000;
+    protected final String clusterUrl;
+    protected final Set<String> scopes;
+    protected final String aadAuthorityUrl;
+
+    TokenProviderBase(@NotNull String clusterUrl, String authorityId) throws URISyntaxException {
+        URI clusterUri = new URI(clusterUrl);
+        this.clusterUrl = String.format("%s://%s", clusterUri.getScheme(), clusterUri.getHost());
+        String scope = String.format("%s/%s", clusterUrl, ".default");
+        scopes = new HashSet<>();
+        scopes.add(scope);
+
+        aadAuthorityUrl = determineAadAuthorityUrl(authorityId);
+    }
+
+    private String determineAadAuthorityUrl(String authorityId) {
+        if (authorityId == null) {
+            authorityId = ORGANIZATION_URI_SUFFIX;
+        }
+
+        String authorityUrl;
+        String aadAuthorityFromEnv = System.getenv("AadAuthorityUri");
+        if (aadAuthorityFromEnv == null) {
+            authorityUrl = String.format("https://login.microsoftonline.com/%s/", authorityId);
+        } else {
+            authorityUrl = String.format("%s%s%s/", aadAuthorityFromEnv, aadAuthorityFromEnv.endsWith("/") ? "" : "/", authorityId);
+        }
+        return authorityUrl;
+    }
+
+    public String acquireAccessToken() throws DataServiceException, DataClientException {
+        IAuthenticationResult accessTokenResult = acquireAccessTokenSilently();
+        if (accessTokenResult == null) {
+            accessTokenResult = acquireNewAccessToken();
+        }
+        return accessTokenResult.accessToken();
+    }
+
+    protected IAuthenticationResult acquireAccessTokenSilently() throws DataServiceException, DataClientException {
+        try {
+            // TODO: MSAL docs say this would throw MsalUiRequiredException, but it doesn't in Java
+            return acquireAccessTokenSilentlyMsal();
+        } catch (MalformedURLException e) {
+            throw new DataClientException(clusterUrl, ERROR_INVALID_AUTHORITY_URL, e);
+        } catch (TimeoutException | ExecutionException e) {
+            return null; // Legitimate outcome, in which case a new token will be acquired
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            return null; // Legitimate outcome, in which case a new token will be acquired
+        }
+    }
+
+    protected IAuthenticationResult acquireAccessTokenSilentlyMsal() throws MalformedURLException, InterruptedException, ExecutionException, TimeoutException, DataServiceException {
+        throw new DataServiceException("Cannot obtain a refresh token for Authentication type '" + this.getClass().getSimpleName() + "'");

```

Comment:
Ha?
refresh token?
Isn't there a way to abstract this away so that it must be implemented by the sub class

Reply:
Good call, we're not getting a refresh token but following a silent flow. Reworded.

We don't want this to be executed by some subclasses. Per the discussion above, if I add another layer of abstraction, I can remove this from the Base. Not sure that's a good idea, though.